### PR TITLE
k256: refactored `lincomb` to use `LinearCombinationExt` trait

### LINKED ISSUE #973: k256: Linear Combanation over Dynamically Allocated Slices
>
              I'd like to add a `_vec` API here too similar to https://github.com/RustCrypto/traits/pull/1376

_Originally posted by @ycscaly in https://github.com/RustCrypto/elliptic-curves/issues/921#issuecomment-1809282828_

Whilst in #921 we decided to expose `lincomb` as-is, I specified that I'd like a similar interface for dynamically-allocated slices. Following that, we did some work that achieved that without code duplication. Here, I suggest to do the same.

### PR DESCRIPTION
**Pull request by ycscaly**
Resolves #973

### CODE REVIEW

**Comment by ycscaly** [comment]
@tarcieri this code is working and looks good to me, only thing is whether we want to expose them as free functions or via a trait (I prefer the second)

**Review by tarcieri — APPROVED** [review]

File: k256/src/arithmetic.rs [hunk 1]
```diff
@@ -12,11 +12,19 @@ pub(crate) mod scalar;
...
+use elliptic_curve::{group, CurveArithmetic};
+
+/// Linear combination.
+///
+/// This trait enables providing an optimized implementation of
+/// linear combinations (e.g. Shamir's Trick).
+pub trait LinearCombination<PointsAndScalars: AsRef<[(Self, Self::Scalar)]>>: group::Curve {
+    /// Calculates `x1 * k1 + ... + xn * kn`.
+    fn linear_combination(points_and_scalars: PointsAndScalars) -> Self;
+}
```
Comment (tarcieri) [inline k256/src/arithmetic.rs:27]: I opened a PR which adds a similar trait upstream to `elliptic-curve` here: https://github.com/RustCrypto/traits/pull/1405

Comment (ycscaly) [reply k256/src/arithmetic.rs]: Ok I adapted the code to work with that

File: k256/src/arithmetic/mul.rs [hunk 2]
```diff
@@ -278,103 +281,83 @@ impl<const D: usize> Default for Radix16Decomposition<D> {
...
-    f: impl Fn(T, S) -> V,
-    x: &[T; N],
-    y: &[S; N],
-    default: V,
-) -> [V; N] {
-    let mut res = [default; N];
-    for i in 0..N {
-        res[i] = f(x[i], y[i]);
+#[cfg(feature = "alloc")]
+impl LinearCombination<Vec<(ProjectivePoint, Scalar)>> for ProjectivePoint {
```
[the remark this answers is not in the document]
Comment (ycscaly) [reply k256/src/arithmetic/mul.rs]: This gives: error[E0277]: the size for values of type `[(projective::ProjectivePoint, arithmetic::scalar::Scalar)]` cannot be known at compilation time

[the remark this answers is not in the document]
Comment (tarcieri) [reply k256/src/arithmetic/mul.rs]: That's merged, so it should work now

File: Cargo.lock [hunk 3]
```diff
@@ -372,7 +372,7 @@ checksum = "7fcaabb2fef8c910e7f4c7ce9f67a1283a1715879a7c230ca9d6d1ae31f16d91"
 [[package]]
 name = "elliptic-curve"
 version = "0.13.6"
-source = "git+https://github.com/RustCrypto/traits.git#4a663341473273c73fd2ca7e37dadd9ba7013fd2"
+source = "git+https://github.com/RustCrypto/traits.git#37507fd8a1314398f9e0d7bbdf25467071b2c72e"
```
Comment (tarcieri) [inline Cargo.lock:375]:
```suggestion
source = "git+https://github.com/RustCrypto/traits.git#723a1a309c4d987ec54c4c1f7bfe3686dc8d5a29"
```

**Review by tarcieri — APPROVED** [review]

File: Cargo.toml [hunk 4]
```diff
@@ -17,5 +17,5 @@ members = [
 [profile.dev]
 opt-level = 2
 
-[patch.crates-io.elliptic-curve]
-git = "https://github.com/RustCrypto/traits.git"
+[patch.crates-io]
+elliptic-curve = {git = "https://github.com/RustCrypto/traits.git"}
```
Comment (tarcieri) [inline Cargo.toml:21]:
```suggestion
[patch.crates-io.elliptic-curve]
git = "https://github.com/RustCrypto/traits.git"
```

**Comment by ycscaly** [comment]
Looks mergable to me

**Comment by tarcieri** [comment]
Will merge when CI is green

**Comment by ycscaly** [comment]
Great thanks @tarcieri now only the multiexp PR is left open on my side https://github.com/RustCrypto/crypto-bigint/pull/248

**Pull request merged**